### ISSUE TITLE
FSDP checkpointing issue

### ISSUE THREAD

**Issue by ananyahjha93** [mosaicml/composer#1703]
FSDP checkpoint is not saved properly.

** Environment **

PyTorch information
-------------------
PyTorch version: 1.12.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.10.4 (main, Mar 31 2022, 08:41:55) [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-125-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM-80GB
GPU 1: NVIDIA A100-SXM-80GB
GPU 2: NVIDIA A100-SXM-80GB
GPU 3: NVIDIA A100-SXM-80GB
GPU 4: NVIDIA A100-SXM-80GB
GPU 5: NVIDIA A100-SXM-80GB
GPU 6: NVIDIA A100-SXM-80GB
GPU 7: NVIDIA A100-SXM-80GB

Nvidia driver version: 495.29.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] pytorch-ranger==0.1.1
[pip3] torch==1.12.1+cu113
[pip3] torch-optimizer==0.3.0
[pip3] torchaudio==0.12.1+cu113
[pip3] torchmetrics==0.7.3
[pip3] torchtext==0.13.1
[pip3] torchvision==0.13.1+cu113
[conda] numpy                     1.23.3                   pypi_0    pypi
[conda] torch                     1.12.1+cu113             pypi_0    pypi
[conda] torchaudio                0.12.1+cu113             pypi_0    pypi
[conda] torchtext                 0.13.1                   pypi_0    pypi
[conda] torchvision               0.13.1+cu113             pypi_0    pypi

Composer information
--------------------
Composer version: 0.11.0
Composer commit hash: None
Host processor model name: AMD EPYC 7352 24-Core Processor
Host processor core count: 48
Number of nodes: 1
Accelerator model name: NVIDIA A100-SXM-80GB
Accelerators per node: 1
CUDA Device Count: 8

** To reproduce

1. train a model using fsdp and save checkpoints
2. reload checkpoint to train the model further
3. loss resets to randomly initialized model levels

## Expected behavior

model should continue training from a loss value where it previously left off. ex - loading from a non-fsdp checkpoint of the same model

loading from a regular checkpoint 

<img width="1051" alt="Screen Shot 2022-11-04 at 3 48 50 AM" src="https://user-images.githubusercontent.com/7491256/199957593-fab460de-a920-4f33-a9b8-9f5d76c5e110.png">

vs loading from an fsdp checkpoint with the same amount of training

<img width="1012" alt="Screen Shot 2022-11-04 at 3 54 12 AM" src="https://user-images.githubusercontent.com/7491256/199957899-6e35477c-2fb4-49c5-a122-2b8f5108c384.png">

**Comment by dblalock** [comment]
@abhi-mosaic @bmosaicml

**Comment by abhi-mosaic** [comment]
Hi @ananyahjha93 ! Could you provide a bit more context into how you are doing the checkpoint loading? Is it:

1) Taking a Composer checkpoint (that may have been trained with DDP or FSDP, that looks like `ep0-ba200.pt`) and loading it directly into the Composer Trainer via `load_path=...`?

2) Taking a Composer checkpoint (that may have been trained with DDP or FSDP, that looks like `ep0-ba200.pt`) and trying to load the model weights into a model class (e.g. HF `GPT2LMHeadModel`) before the model is passed to the Trainer?

I will do a quick sweep on 1) with the 2x2 matrix of saving + loading on either FSDP or DDP. But my gut feeling is that we have tests for this, and I've also been reloading FSDP runs via this way and the loss resumes at the expected place.

For 2), it would be really useful if you could send a code snippet for how you are doing the model resumption. In our expereince there can be a lot of silent errors when resuming HF models with Composer checkpoints, because you have to extract the `state_dict['state']['model']` and also remove any prefixes like `model.` from the state dict before reloading. Normally, most PyTorch modules would throw an error if something goes wrong, but HF models just print a warning and continue, and that warning that can easily get lost in the logs.

Here's the script I use whenever I want to extract just the HF model weights from a Composer checkpoint:
```python
import torch
import sys

path = sys.argv[1]
out_path = '/'.join(path.split('/')[:-1] + ['pytorch_model.bin'])
print (out_path)

composer_ckpt = torch.load(path, map_location=torch.device('cpu'))
state_dict = composer_ckpt['state']['model']
torch.nn.modules.utils.consume_prefix_in_state_dict_if_present(state_dict, prefix='model.')
torch.save(state_dict, out_path)
```

And to confirm whether the HF model has correctly reloaded, I should only see 1 or 2 missing weights printed by HF:
```
[INFO|modeling_utils.py:2155] 2022-11-04 18:10:42,189 >> loading weights file 3b-50k/pytorch_model.bin
[WARNING|modeling_utils.py:2598] 2022-11-04 18:11:19,264 >> Some weights of the model checkpoint at 3b-50k/ were not used when initializing GPT2ForMultipleChoice: ['lm_head.weight']
```
In this case, I'm loading a pretrained `GPT2LMHeadModel` into a downstream `GPT2ForMultipleChoice` and so it makes sense that the `lm_head` from pretraining does not get loaded. But all the other weights are loaded correctly.

Let me know if this helps! And please send any relevant code snippets so we can understand how you are loading the checkpoint.
